oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStore.java

```diff
@@ -1411,7 +1411,7 @@ public <T extends Document> void prefetch(Collection<T> collection,
         List<String> resultKeys = new ArrayList<>(keys.size());
         CacheChangesTracker tracker = null;
         if (collection == Collection.NODES) {
-            tracker = nodesCache.registerTracker(keys);
+            tracker = nodesCache.registerTracker(new HashSet<>(keys));
```

Comment:
Is this to remove duplicates? Using HashSet will re-order the entries, but I guess that's fine. 

Possibly it makes sense to sort the entries so that MongoDB can retrieve them faster (a MongoDB specific optimization); I could imagine that MongoDB doesn't do that itself. Or so that compression over the wire can work better (I _think_ LZ4 is used?)

Reply:
> Is this to remove duplicates?

No, it just registers the given keys with a `CacheChangesTracker`. Order is not important here. The `CacheChangesTracker` just needs an efficient lookup whether a key is in this set.

> Possibly it makes sense to sort the entries so that MongoDB can retrieve them faster

I don't know if that helps and I would only sort if there is evidence that it is faster.

> I think LZ4 is used?

It _can_ be used. Compression over the wire is a connection [option](https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/connection/network-compression/#specify-compression-algorithms).

Reply:
> it just registers the given keys

Yes, but we had "registerTracker(keys)" before and you changed it to "registerTracker(new HashSet<>(keys))" and I wanted to understand the reason for this change

```
-            tracker = nodesCache.registerTracker(keys);
+            tracker = nodesCache.registerTracker(new HashSet<>(keys));

```

> I would only sort if there is evidence that it is faster.

Sure.

> Compression over the wire is a connection option

Thanks! So no LZ4, but there's Snappy. (which is a bit slower than LZ4 I think)

Reply:
Please ignore my comments about compression and ordering... I see now this doesn't apply here.

But "keys" is already "Set<String> keys = new HashSet<>();", and you wrapped that inside a new HashSet, what is the reason?

Reply:
Ah, is it because "Keys collection passed to registerTracker must not be modified later"  and you wanted to ensure the set is not modified later? I think it's fine... It might make sense to add a comment in the code

Reply:
Yes, that's exactly the reason. Good point about the comment. I'll add that.